### PR TITLE
[12.0][IMP] web_responsive: Document Viewer

### PR DESCRIPTION
**Pull request by Tardo**
document_viewer

cc @tecnativa TT21158

### CODE REVIEW

File: web_responsive/static/src/css/web_responsive.scss [hunk 1]
```diff
@@ -576,3 +578,83 @@ html .o_web_client .o_main .o_main_content {
 .oe_wait {
   cursor: progress;
 }
+
+// Document Viewer
+.o_web_client.o_chatter_position_sided {
+    .o_document_viewer {
+        z-index: 10 !important;
```
Comment (pedrobaeza) [inline web_responsive/static/src/css/web_responsive.scss]: Can you avoid somehow the `!important` part for not forcing so many styles?

**Review by pedrobaeza — COMMENTED** [review]
Excellent new feature!

We need to document it on the README file as the other ones, and the GIF is perfect for being included. You can put this text:

```
* When the chatter is configured on the side part, the document viewer fills that 
  part for side-by-side reading instead of full screen. You can still put it on full 
  width preview clicking on the new maximize button.
```

File: web_responsive/static/src/css/web_responsive.scss (hunk 1 above)
Comment (yajo) [reply web_responsive/static/src/css/web_responsive.scss]: Usually the way to do this is to use a more specific selector, or use the same selector but later.

BTW please add some comments to understand what do these things do.

File: web_responsive/static/src/js/web_responsive.js [hunk 2]
```diff
@@ -525,4 +526,35 @@ odoo.define('web_responsive', function (require) {
...
+            } else {
+                this.$el.addClass('o_responsive_document_viewer');
+                $target.addClass('o_maximize_btn')
+                    .removeClass('o_minimize_btn');
+                $target.find('.fa-window-minimize')
+                    .addClass('fa-window-maximize')
+                    .removeClass('fa-window-minimize')
+                    .attr('title', _("Maximize"))
+                    .attr('aria-label', _("Maximize"));
+            }
```
Comment (yajo) [inline web_responsive/static/src/js/web_responsive.js]: :thinking: Have you considered adding both buttons but leaving one hidden by default, and just toggle their visibility? I think it will simplify code quite a bit... Well, actually you could even show/hide them with just CSS probably.

File: web_responsive/static/src/xml/document_viewer.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            <attribute name="class">modal o_modal_fullscreen o_document_viewer o_responsive_document_viewer</attribute>
+            <attribute name="data-backdrop">false</attribute>
+        </t>
+    </t>
+
+    <t t-extend="DocumentViewer.Content">
+        <t t-jquery=".o_close_btn" t-operation="replace">
+            <div class="o_buttons float-right mr8">
+                <a role="button" class="mr8 o_maximize_btn d-none" href="#" aria-label="Maximize" title="Maximize"><i class="fa fa-window-maximize"></i></a>
+                <a role="button" class="o_close_btn" href="#" aria-label="Close" title="Close"><i class="fa fa-close"></i></a>
```
Comment (yajo) [inline web_responsive/static/src/xml/document_viewer.xml]: I know adding `href="#"` is a common pattern, but usually it breaks your browsing history unless explicitly calling `event.preventDefault()`. Remove that if it's not needed please.

File: web_responsive/views/web.xml [hunk 4]
```diff
@@ -11,8 +11,8 @@
               inherit_id="web.webclient_bootstrap"
               name="App Drawer - Web Client"
               >
-        <xpath expr="//*[hasclass('o_main')]" position="attributes">
-            <attribute name="t-attf-class">o_main o_chatter_position_{{ request.env.user.chatter_position or 'normal' }}</attribute>
+        <xpath expr="//t[@t-set='body_classname']" position="attributes">
+            <attribute name="t-value">'o_web_client o_chatter_position_' + (request.env.user.chatter_position or 'normal')</attribute>
```
Comment (yajo) [inline web_responsive/views/web.xml:15]: Why this change? It's a bit weird.

File: web_responsive/static/src/css/web_responsive.scss [hunk 5]
```diff
@@ -576,3 +578,83 @@ html .o_web_client .o_main .o_main_content {
...
+            }
+        }
+    }
+}
+.o_web_client:not(.o_chatter_position_sided) {
+    .o_split_pdf_area {
+        display: inline !important;
+    }
+}
+//-- Generic
```
Comment (yajo) [inline web_responsive/static/src/css/web_responsive.scss]:
```suggestion

// Generic
```

BTW what does Generic mean? :thinking:

**Review by yajo — CHANGES_REQUESTED** [review]
This needs comments.

Nice feature OTOH!

File: web_responsive/views/web.xml (hunk 4 above)
Comment (Tardo) [reply web_responsive/views/web.xml:15]: This change was made because 'o_document_viewer' is out from '.o_main'.

**Comment by Tardo** [comment]
Changes done! Thanks for the reviews... and thanks @pedrobaeza for the original idea!

**Review by pedrobaeza — APPROVED** [review]
OK by my side.

**Comment by legalsylvain** [comment]
Hi @tardo. Thanks a lot for this feature. looks very good ! 
I'd like to test, but runbot link is not available and there are no link here : https://runbot.odoo-community.org/
@gurneyalex : An idea ? (after your vacation, of course ;-))

**Comment by pedrobaeza** [comment]
After the runbot upgrade, seems that the CI hook that links in the pull requests the runbot is not working. You can browse them manually through https://runbot.odoo-community.org/runbot (and selecting the repo in the dropdown). For this, I save your time providing the direct link ;):

https://runbot.odoo-community.org/runbot/build/3403398

File: web_responsive/static/src/js/web_responsive.js [hunk 6]
```diff
@@ -525,4 +526,29 @@ odoo.define('web_responsive', function (require) {
...
+        start: function () {
+            this.$btnMaximize = this.$('.o_maximize_btn');
+            this.$btnMinimize = this.$('.o_minimize_btn');
+            return this._super.apply(this, arguments);
+        },
+
+        _onClickMaximize: function () {
+            this.$el.removeClass('o_responsive_document_viewer');
+            this.$btnMaximize.addClass('d-none');
+            this.$btnMinimize.removeClass('d-none');
```
Comment (yajo) [inline web_responsive/static/src/js/web_responsive.js]: Can't this be done by CSS? :thinking: 

```scss
.o_responsive_document_viewer {
    .o_maximize_btn {
      display: none;
    }
}
// ... etc.
```

**Review by yajo — APPROVED** [review]
Cool now. Just one suggestion to consider.

**Comment by Tardo** [comment]
Thanks @Yajo changes done!

**Comment by pedrobaeza** [comment]
/ocabot merge major

Now please migrate it to v13
